reverb/cc/sampler.cc

```diff
@@ -208,6 +208,13 @@ class GrpcSamplerWorker : public SamplerWorker {
       }
       context_ = std::make_unique<grpc::ClientContext>();
       context_->set_wait_for_ready(false);
+
+      // Buffer time to account for connection overhead
+      constexpr auto CONNECTION_BUFFER_TIME = std::chrono::milliseconds(200);
+
```

Comment:
The deadline must only be set if the rate_limiter_timeout is set. Also actually it isn't quite this easy I just realised since you also can't set the deadline when using the datasets (which is the recommended way).

Furthermore, there is a whole bunch of tests needed here to verify that all the behaviours are as expected.

Reply:
You are right, I oversight that `if-clause` to check if the `rate_limiter_timeout` is **set** or **not**.

When utilizing the **“datasets”** method. It implies that while using **“datasets”**, we cannot set a deadline for the RPC call as it might be getting transferred. Am I getting this right?

Reply:
Yes the normal behaviour is to wait for the connection to be establish and only break out if the rate limiter has blocked for the specified amount of time.

Basically I think there might be a need to clearly separate the concepts here and follow a pattern more similar to the writers. I.e. instead of just returning a generator from `client.sample` we might have to return a special object which has support for getting the next item with an optional deadline.

Really though, this feature will be quite a lot of work and I can't guarantee that I'll be able to accept it unless it is both very cleanly implemented and guaranteed to work in all cases. How important is this feature for you? Unless it is absolutely critical I would probably recommend you not to spend more time on this.

Reply:
I am just looking for my learning and to improve my programming ability. Since 2017 I did ML/AI stuff but left open source without even trying. Now I am doing what I left and seeking places where I can learn better with some guidance. 

So, this is all for my learning. =)